lib/src/package_analysis/common.dart

```diff
@@ -1,12 +1,204 @@
+  required bool wipeTarget,
+}) async {
+  // delete the target directory, if it exists and wipe is enabled
+  if (wipeTarget && await Directory(destination).exists()) {
+    await Directory(destination).delete(recursive: true);
+  }
+
+  // construct pubspec for the pointer package
+  final pointerPubspec = <String, dynamic>{
+    'name': 'dummy_package',
+    'environment': {
+      'sdk': '>=2.12.0 <3.0.0',
+    },
+    'dependencies': {},
+  };
+  pointerPubspec['dependencies'][name] = version;
+
+  // if pubspec.yaml exists, delete it and create an empty file
+  final pubspecFile = File(path.join(destination, 'pubspec.yaml'));
+  if (await pubspecFile.exists()) {
+    await pubspecFile.delete();
+  }
+  await pubspecFile.create(recursive: true);
+
+  // write pubspec to disk
+  await pubspecFile.writeAsString(json.encode(pointerPubspec));
+
+  // fetch dependencies (the only non-transitive dependency is [name])
+  await fetchDependencies(destination);
+}
+
+/// Fetches dependencies in [destination].
+Future<void> fetchDependencies(String destination) async {
+  final result = await Process.run(
+    Platform.resolvedExecutable,
+    ['pub', 'get'],
+    workingDirectory: destination,
+  );
+  if (result.exitCode != 0) {
+    throw ProcessException(
+      Platform.resolvedExecutable,
+      ['pub', 'get'],
+      result.stderr.toString(),
+      result.exitCode,
+    );
+  }
+}
+
+/// Given the location of a package and the name of one of its dependencies,
+/// return the path of the dependency, or null if it cannot be resolved.
+/// Ensure that the dependencies of the target package are fetched.
+Future<String?> getDependencyDirectory(
+  PackageAnalysisContext packageAnalysisContext,
+  String packageLocation,
+  String dependencyName,
+) async {
+  final dependencyUri = Uri.parse('package:$dependencyName/');
+  final dependencyFilePath = packageAnalysisContext.uriToPath(dependencyUri);
+  print('path for $dependencyName in $packageLocation is $dependencyFilePath');
+  return dependencyFilePath == null ? null : path.dirname(dependencyFilePath);
+}
+
+/// Recursively fetch all the files in the given folder
+List<resource.File> getAllFiles(resource.Folder folder) {
+  final files = <resource.File>[];
+  for (final child in folder.getChildren()) {
+    if (child is resource.File) {
+      files.add(child);
+    } else if (child is resource.Folder) {
+      files.addAll(getAllFiles(child));
+    } else {
+      throw StateError(
+          'Failed to determine subclass of Resource ${child.path}.');
+    }
+  }
+  return files;
+}
+
+/// Fetch all the hosted dependencies used by a package at a given location
+Future<Map<String, HostedDependency>> getHostedDependencies(
+    String targetPackageDir) async {
+  // fetch a Map of all dependencies from the target package folder
+  final allDependencies = Pubspec.parseYaml(
+          await File(path.join(targetPackageDir, 'pubspec.yaml'))
+              .readAsString())
+      .dependencies;
```

Comment:
Just an idea... Because FormatException and IOException should never happen here, if they do then it's because we gave a path that's not really a package.
Handling them and throwing a generic StateError is nice because we don't want someone calling  `getHostedDependencies` to handle `FormatException` or `IOException`, someone calling `getHostedDependencies` should always ensure they are pointing to a real package.

```suggestion
try {
  final allDependencies = Pubspec.parseYaml(
          await File(path.join(targetPackageDir, 'pubspec.yaml'))
              .readAsString())
      .dependencies;
} on FormatException {
  throw StateError('Failed to parse pubspec.yaml from "$targetPackageDir" check that it is a valid package');
} on IOException {
  throw StateError('Failed to read pubspec.yaml from "$targetPackageDir" check that it is a valid package');
}
```

Reply:
I don't think this error handling is necessary

Pubspec.parseYaml() is invoked in three places so far - in two of them the relevant pubspec file is generated beforehand, and in the last its presence is verified (before the summary is generated):
```dart
if (!await File(path.join(packageLocation, 'pubspec.yaml')).exists()) {
  throw StateError('The target directory must contain a package.');
}
```